pipeline-model-api/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/ast/AbstractModelASTParallelContent.java

```diff
@@ -0,0 +1,244 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.pipeline.modeldefinition.ast;
+
+import net.sf.json.JSONObject;
+import org.jenkinsci.plugins.pipeline.modeldefinition.validator.ModelValidator;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Represents an individual Stage and the {@link ModelASTBranch}s it may contain.
+ *
+ * @author Kohsuke Kawaguchi
+ * @author Andrew Bayer
+ * @see ModelASTPipelineDef
+ */
+public abstract class AbstractModelASTParallelContent extends ModelASTElement {
```

Comment:
Bad name, it indicates that classes extending this will have parallel content. But the actual parallel stuff is implemented someplace else.
Should be some thing like `AbstractASTPseudoStage` or something.

Reply:
Yeah, I am emphatically not in love with the name. =)

Reply:
...fwiw, this will probably get renamed to `ModelASTAbstractStage` or something like that, as we generalize this so that you can actually do a top-level `stage` with a group within it.

Reply:
Actually, this gets stupid simpler once we just make it so that a `stage` can contain one of `steps`, `parallel`, or `stages` (or whatever we rename `stages` to in this context), since then it's just one class anyway.